### PR TITLE
Yet another stab at fixing ASTAR

### PR DESCRIPTION
**Pull request by MadmanMartian**
Adds an ASTAR pathing sub

Adds a little crab bot that Astar paths to wherever you click on the screen

Gunks up ASTAR with a bunch of debug (colouring tiles that are processed)

~~Currently ASTAR tends to overprocess when it is looking at diagonals, to the point of even going the opposite direction in some cases.~~

- [x] Fix ASTAR so it doesn't total the server when it has to process a diagonal path
- [x] Make an ASTAR path making sub so we don't have to deal with a while loop
 - [ ] Convert existing bots over to the ASTAR path making and path moving subs

### CODE REVIEW

**Comment by SonixApache** [comment]
cute debug bot, don't delete it when you're done, would be fun for bus (minus the world chat messages)

also doing gods work, good luck

File: code/defines/procs/AStar.dm [hunk 1]
```diff
@@ -343,3 +352,18 @@ proc/AStar(start,end,adjacent,dist,maxnodes,maxnodedepth = 30,mintargetdist,minn
 	return 0
 
 /////////////////////////////////////////////////////////////////////////
+
+/atom/proc/make_astar_path(var/atom/target)
```
Comment (jknpj) [inline code/defines/procs/AStar.dm]: RTS spessmen soon, fellow stalkers

**Comment by MadmanMartian** [comment]
Diagnosis of the previous issue:

It was checking if any new nodes were closer to the start than others (???) then adding that to the currently held nodes proximity to the start, then checking if that value was less than the currently held nodes proximity to the start (This is impossible)

So, you'd have it looking in a circle around the start point to try and find the path, rather than sending tendrils outwards to try and find a way to the end, creeping towards the end location should the distance be lower to the end than elsewhere.

**Comment by MadmanMartian** [comment]
For the path maker, may need to port callback

 - Thing requests path
 - new astar construction datum is created and registered
 - astar pathmaker processes a step of it every subtick
 - Thing requests a path. If its to and from the same destination it is told to wait, otherwise the previous astar construction datum is dropped
 - once the path is complete, the path construction datum calls back to the requester with the path list, then self terminates

File: code/defines/procs/AStar.dm [hunk 2]
```diff
@@ -118,30 +118,30 @@ length to avoid portals or something i guess?? Not that they're counted right no
 /PathNode
 	var/turf/source //turf associated with the PathNode
 	var/PathNode/prevNode //link to the parent PathNode
-	var/f		//A* Node weight (f = g + h)
-	var/g		//A* movement cost variable
-	var/h		//A* heuristic variable
-	var/nt		//count the number of Nodes traversed
+	var/total_node_cost		//A* Node weight (f = g + h)
```
Comment (SonixApache) [inline code/defines/procs/AStar.dm]: if you're gonna change the var names, change them in the comments as well

Comment (Zth--) [reply code/defines/procs/AStar.dm]: or even better: remove the comments

**Comment by MadmanMartian** [comment]
Once this is done, reminder to self: Fix doors CanAstarPass not accounting for if the airlock is powered

**Comment by MadmanMartian** [comment]
Making it so patrol pathong is part of the base bot, removing a lot of copypaste from the other bots

**Comment by Kurfursten** [comment]
Let us know if you ever do anything with this again

**Comment by MadmanMartian** [comment]
sorry, I have been a bit busy of late. I can get onto this now though.

It's just a matter of rewriting most of the bots

**Comment by Kurfursten** [comment]
Feel free to open this again yourself if you continue work on it

**Pull request closed without merging**